### PR TITLE
feat: add wallpaper provider

### PR DESCRIPTION
**Pull request by zorowk**

### CODE REVIEW

File: examples/tinywl-new/qmlengine.cpp [hunk 1]
```diff
@@ -63,3 +63,13 @@ QQuickItem *QmlEngine::createTaskBar(Output *output, QQuickItem *parent)
 
     return item;
 }
+
+WallpaperImageProvider *QmlEngine::wallpaperImageProvider()
+{
+    if (!wallpaperProvider && !this->imageProvider("wallpaper")) {
```
Comment (zccrs) [inline examples/tinywl-new/qmlengine.cpp:69]: Should assert `!this->imageProvider("wallpaper")`

File: examples/tinywl-new/wallpaperimage.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
+// Copyright (C) 2024 WenHao Peng <pengwenhao@uniontech.com>.
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:1]: // Copyright (C) 2024 UnionTech Software Technology Co., Ltd.

File: examples/tinywl-new/wallpaperimage.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "wallpaperprovider.h"
+#include "wallpaperimage.h"
+#include "helper.h"
+
+WallpaperImage::WallpaperImage(QQuickItem *parent)
+    :QQuickImage(parent)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:9]: `: QQuickImage`

File: examples/tinywl-new/wallpaperimage.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+{
+
+}
+
+QString WallpaperImage::userId()
+{
+    return m_userId;
+}
+
+void WallpaperImage::setUserId(const QString& name)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:25]: Is it a `pid`? pid's type should is int, if is user name, this function name should is `setUserName`

File: examples/tinywl-new/wallpaperimage.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        updateSource();
+    }
+}
+
+QString WallpaperImage::screenId()
+{
+    return m_screenId;
+}
+
+void WallpaperImage::setScreenId(const QString& screen_id)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:39]: Like as `setUserId`, What's screen id? xdg output name?
`void setOutput(Output *output)` is better.

File: examples/tinywl-new/wallpaperimage.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        updateSource();
+    }
+}
+
+QString WallpaperImage::fileName()
+{
+    return m_fileName;
+}
+
+void WallpaperImage::setFileName(const QString& path)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:53]: No, don't allows set image file.

File: examples/tinywl-new/wallpaperimage.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        updateSource();
+    }
+}
+
+int WallpaperImage::workspaceId()
+{
+    return m_workspaceId;
+}
+
+void WallpaperImage::setWorkspaceId(const int workspace_id)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:67]: `setWorkspace(WorkspaceContainer *workspace)`

File: examples/tinywl-new/wallpaperprovider.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+#include "wallpaperprovider.h"
+
+#include <QDir>
+#include <QFileInfo>
+#include <QQuickWindow>
+#include <QImageReader>
+#include <QSGTexture>
+
+WallpaperTextureFactory::WallpaperTextureFactory(WallpaperImageProvider* provider, const QImage &image)
+    :m_wallpaperProvider(provider)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperprovider.cpp:13]: `: `

File: examples/tinywl-new/wallpaperprovider.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        || image.format() == QImage::Format_RGB32) {
+        im = image;
+    } else {
+        im = image.convertToFormat(QImage::Format_ARGB32_Premultiplied);
+    }
+    size = im.size();
+}
+
+WallpaperTextureFactory::WallpaperTextureFactory(WallpaperImageProvider *provider, const QString &id)
+    :m_wallpaperProvider(provider)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperprovider.cpp:25]: `: `

File: examples/tinywl-new/wallpaperprovider.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    ,m_wallpaperId(id)
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperprovider.cpp:26]: ", "

File: examples/tinywl-new/wallpaperprovider.cpp [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+QSGTexture *WallpaperTextureFactory::createTexture(QQuickWindow *window) const
+{
+    if (m_textureExist)
+        m_texture = m_wallpaperProvider->getExistTexture(m_wallpaperId);
+
+    if (!m_texture)
+        m_texture = window->createTextureFromImage(im, QQuickWindow::TextureCanUseAtlas);
+
+    static bool transient = qEnvironmentVariableIsSet("QSG_TRANSIENT_IMAGES");
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperprovider.cpp:39]: Don't check this env, always clear image here

File: examples/tinywl-new/wallpaperprovider.cpp [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+QQuickTextureFactory *WallpaperImageProvider::requestTexture(const QString &id, QSize *size, const QSize &requestedSize)
+{
+    QQuickTextureFactory *factory = nullptr;
+    QSize readSize;
+
+    if (textureCache.contains(id)) {
+        auto cache_factory = textureCache[id];
+        readSize = cache_factory->textureSize();
+        if (requestedSize.width() * requestedSize.height() < readSize.width() * readSize.height()) {
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperprovider.cpp:84]: Why is `width * height`?

`if (requestedSize.width() < readSize.width() && requestedSize.height() < readSize.height())`

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl-new/wallpaperprovider.cpp (hunk 12 above)
Comment (zorowk) [reply examples/tinywl-new/wallpaperprovider.cpp]: i change

File: examples/tinywl-new/PrimaryOutput.qml [hunk 13]
```diff
@@ -77,11 +77,16 @@ OutputItem {
         }
     }
 
-    Image {
+    // /home/uos/.cache/treeland/wallpaper/1000/x11-x/0/deepin-theme-1.jpg
```
Comment (zccrs) [inline examples/tinywl-new/PrimaryOutput.qml:80]: Remove

File: examples/tinywl-new/PrimaryOutput.qml [hunk 14]
```diff
@@ -77,11 +77,16 @@ OutputItem {
...
         id: background
-        source: "file:///usr/share/wallpapers/deepin/desktop.jpg"
-        fillMode: Image.PreserveAspectCrop
+        userId: 1000
+        output: rootOutputItem.output
+
+        fillMode: Image.Tile
+        cache: false
         asynchronous: true
+        sourceSize: Qt.size(1920, 1080)
```
Comment (zccrs) [inline examples/tinywl-new/PrimaryOutput.qml:89]: Why is 1920 1080?

File: examples/tinywl-new/wallpaperimage.cpp [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+#include <QDir>
+
+WallpaperImage::WallpaperImage(QQuickItem *parent)
+    : QQuickImage(parent)
+{
+    auto provider = Helper::instance()->qmlEngine()->wallpaperImageProvider();
+    connect(provider, &WallpaperImageProvider::wallpaperTextureUpdate, this, &WallpaperImage::updateWallpaperTexture);
+
+    auto workspace = Helper::instance()->workspace();
+    connect(workspace, &Workspace::currentChanged, this, &WallpaperImage::updateSource);
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:20]: No, don't care current workspace here, the current workspace set by qml, for an example:

````
WallpaperImage {
    ...
    workspace: Helper.workspace.current
}
````

File: examples/tinywl-new/wallpaperimage.cpp [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+void WallpaperImage::updateSource()
+{
+    auto workspace = Helper::instance()->workspace();
+    if (m_userId == -1 ||
+        !m_output ||
+        !workspace) {
+        return;
+    }
+
+    QString home = QStandardPaths::writableLocation(QStandardPaths::HomeLocation);
```
Comment (zccrs) [inline examples/tinywl-new/wallpaperimage.cpp:65]: WallpaperImage don't read local file, always use `image://wallpaper/xxxx` url to get wallpaper texture from WallpaperProvider.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl-new/PrimaryOutput.qml (hunk 14 above)
Comment (zorowk) [reply examples/tinywl-new/PrimaryOutput.qml:89]: This is the value of the requestsize of the image provider. If not set, it is (-1, -1)

**Pull request merged**